style/infra-style.md

```diff
@@ -0,0 +1,109 @@
+
+It can be our _own_ code, written earlier, that doesn't make sense.
+
+From https://www.techtarget.com/searchsoftwarequality/feature/Follow-Googles-lead-with-programming-style-guides:
+
+When code is consistent and readable across the organization, Shaindel Schwartz advises, engineers see these benefits:
+- ability to focus on what's getting done, rather than the minutiae of how it's coded;
+- quick understanding of any codebase through familiar interfaces and formats;
+- simpler scaling and implementation of automation; and
+- easier change, such as product ownership transfers, new personnel onboarding and code handoffs.
+
+
+## Our Goals:
+- improve maintainability
+- reduce cognitive load when reading code; we grok it faster if choices are consistent across codebases
+- avoid mistakes, e.g. typos
+- continually improve code
+- follow best practices in broader community
+- avoid disagreements about style
+
+- do as much as possible with automated tools
+
+## Tools
+
+Ruby:
+  - [rubocop](https://rubocop.org/) [cop documentation](https://docs.rubocop.org/rubocop) [github](https://github.com/rubocop/rubocop)
+  - [rubocop-rspec cop documentation](https://docs.rubocop.org/rubocop-rspec/cops.html) [github](https://github.com/rubocop/rubocop-rspec)
+  - [rubocop-rails cop documentation](https://docs.rubocop.org/rubocop-rails/cops.html) [github](https://github.com/rubocop/rubocop-rails)
+  - (not often used by us)[rubocop-performance cop documentation](https://docs.rubocop.org/rubocop-performance/cops_performance.html) [github](https://github.com/rubocop/rubocop-performance)
+
+We want these tools to be exercised when a pull request is created, at a minimum.
+
+
+## Editing existing codebases
+
+Most of our work is editing existing code bases, whether they are 3 months old or 15 years old.
+
+### RULE:  Follow the existing rules.  Strive for no new exceptions.
+
+### Should it be a goal to reduce .rubocop_todo.yml to empty?
+
+(TBD)
+
+
+## Starting a New Code Base
+
+Start with default style guides + exceptions below.
+
+
+## When Style exceptions are acceptable
+
+For long classes/methods/blocks, consider SOLID:
+
+    **S**ingle responsibility
+    **O**pen-Closed (open for extension, closed for modification)
+    **L**iskov substitution
+    **I**nterface segregation (clients should not be forced to depend of interfaces they don't use)
+    **D**ependency inversion (depend on abstractions, not concretions)
+
+However, refactoring can reduce readability:
+- the reader now has to jump around more in the code without a clear gain in readability
+- method/class names matter!
+
+### When can we turn a rule off?
+
+(TBD)
+
+
+## What to do when keeping an exception
+
+- Prefer an inline exception to one in .rubocop.yml
+   - On same line if it's a single line to be excepted and one rule
+   - Above disable, below disable if multiple rules or multiple lines excepted
+- Consider a comment indicating why ("shameless green", "we need UTF-8 here")
```

Comment:
Is "shameless green" really a helpful comment?

Reply:
@justinlittman it says "I got it working and I know this code could be better" - kind of similar to "TODO: improve this"

Reply:
I'm on the fence about the value of adding comments for in-line Rubocop exceptions, entirely, but I'm open to alternative opinions.

Reply:
Yeah, the "shameless green" terminology would be lost on me. I would think for in-line exceptions, this would be up to the dev/reviewer. For instance, a long line is pretty obvious but maybe complexity is unavoidable for a reason that could be clarified in a comment.

Reply:
Are there examples of where we've included comments on inline exceptions or would this be a new practice?

Reply:
This is now marked TBD.